File: python/109. Convert Sorted List to Binary Search Tree.py
#coding=utf-8
__author__ = 'xuxuan'
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def sortedListToBST(self, head):
        """
        :type head: ListNode
        :rtype: TreeNode
        """
        if head is None:
            return None
        fast=head.next
        slow=head
        end=None
        while fast and fast.next :
            fast=fast.next.next
            end=slow
            slow=slow.next
        ans=TreeNode(slow.val)
        ans.right=self.sortedListToBST(slow.next)
        if end:
            end.next=None
            ans.left=self.sortedListToBST(head)
        return ans

File: java/109. Convert Sorted List to Binary Search Tree.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        return f(head,null);
    }
    public TreeNode f(ListNode head,ListNode end) {
        if (head==end) {
            return null;
        }
        ListNode fast=head;
        ListNode slow=head;
        while (fast.next!=end&&fast.next.next!=end) {
            fast=fast.next.next;
            slow=slow.next;
        }
        TreeNode node=new TreeNode(slow.val);
        node.right=f(slow.next,end);
        node.left=f(head,slow);
        return node;
    }
}